File: src/main/java/com/company/apimgmt/repository/RelationshipRepository.java
package com.company.apimgmt.repository;

import com.company.apimgmt.entity.RelationshipEntity;
import com.company.apimgmt.enums.AuthType;
import com.company.apimgmt.enums.EntityType;
import io.smallrye.mutiny.Uni;
import io.vertx.core.json.JsonObject;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import io.vertx.mutiny.pgclient.PgPool;
import io.vertx.mutiny.sqlclient.Row;
import io.vertx.mutiny.sqlclient.RowSet;
import io.vertx.mutiny.sqlclient.Tuple;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@ApplicationScoped
public class RelationshipRepository {
    
    @Inject
    PgPool client;
    
    public Uni<RelationshipEntity> create(RelationshipEntity relationship) {
        return client.preparedQuery(
            "INSERT INTO relationships (id, caller_type, caller_id, callee_type, callee_id, " +
            "endpoint_id, auth_type, auth_config, description, created_at, updated_at) " +
            "VALUES ($1, $2, $3, $4, $5, $6, $7, $8::jsonb, $9, $10, $11) RETURNING *")
            .execute(Tuple.from(java.util.Arrays.asList(
                relationship.id,
                relationship.callerType.name(),
                relationship.callerId,
                relationship.calleeType.name(),
                relationship.calleeId,
                relationship.endpointId,
                relationship.authType != null ? relationship.authType.name() : null,
                relationship.authConfig != null ? relationship.authConfig.encode() : null,
                relationship.description,
                relationship.createdAt,
                relationship.updatedAt
            )))
            .onItem().transform(rows -> mapToRelationshipEntity(rows.iterator().next()));
    }
    
    public Uni<RelationshipEntity> findById(UUID id) {
        return client.preparedQuery("SELECT * FROM relationships WHERE id = $1")
            .execute(Tuple.of(id))
            .onItem().transform(rows -> {
                var iterator = rows.iterator();
                return iterator.hasNext() ? mapToRelationshipEntity(iterator.next()) : null;
            });
    }
    
    public Uni<List<RelationshipEntity>> findAll() {
        return client.query("SELECT * FROM relationships ORDER BY created_at DESC")
            .execute()
            .onItem().transform(this::mapToRelationshipEntityList);
    }
    
    public Uni<List<RelationshipEntity>> findByCaller(EntityType type, UUID id) {
        return client.preparedQuery(
            "SELECT * FROM relationships WHERE caller_type = $1 AND caller_id = $2 ORDER BY created_at DESC")
            .execute(Tuple.of(type.name(), id))
            .onItem().transform(this::mapToRelationshipEntityList);
    }
    
    public Uni<List<RelationshipEntity>> findByCallee(EntityType type, UUID id) {
        return client.preparedQuery(
            "SELECT * FROM relationships WHERE callee_type = $1 AND callee_id = $2 ORDER BY created_at DESC")
            .execute(Tuple.of(type.name(), id))
            .onItem().transform(this::mapToRelationshipEntityList);
    }
    
    public Uni<RelationshipEntity> update(RelationshipEntity relationship) {
        return client.preparedQuery(
            "UPDATE relationships SET caller_type = $1, caller_id = $2, callee_type = $3, " +
            "callee_id = $4, endpoint_id = $5, auth_type = $6, auth_config = $7::jsonb, " +
            "description = $8, updated_at = $9 WHERE id = $10 RETURNING *")
            .execute(Tuple.from(java.util.Arrays.asList(
                relationship.callerType.name(),
                relationship.callerId,
                relationship.calleeType.name(),
                relationship.calleeId,
                relationship.endpointId,
                relationship.authType != null ? relationship.authType.name() : null,
                relationship.authConfig != null ? relationship.authConfig.encode() : null,
                relationship.description,
                relationship.updatedAt,
                relationship.id
            )))
            .onItem().transform(rows -> {
                var iterator = rows.iterator();
                return iterator.hasNext() ? mapToRelationshipEntity(iterator.next()) : null;
            });
    }
    
    public Uni<Boolean> delete(UUID id) {
        return client.preparedQuery("DELETE FROM relationships WHERE id = $1")
            .execute(Tuple.of(id))
            .onItem().transform(rows -> rows.rowCount() > 0);
    }
    
    private RelationshipEntity mapToRelationshipEntity(Row row) {
        RelationshipEntity entity = new RelationshipEntity();
        entity.id = row.getUUID("id");
        entity.callerType = EntityType.valueOf(row.getString("caller_type"));
        entity.callerId = row.getUUID("caller_id");
        entity.calleeType = EntityType.valueOf(row.getString("callee_type"));
        entity.calleeId = row.getUUID("callee_id");
        entity.endpointId = row.getUUID("endpoint_id");
        
        String authTypeStr = row.getString("auth_type");
        entity.authType = authTypeStr != null ? AuthType.valueOf(authTypeStr) : null;
        
        Object authConfigObj = row.getValue("auth_config");
        if (authConfigObj != null) {
            entity.authConfig = new JsonObject(authConfigObj.toString());
        }
        
        entity.description = row.getString("description");
        entity.createdAt = row.getLocalDateTime("created_at");
        entity.updatedAt = row.getLocalDateTime("updated_at");
        
        return entity;
    }
    
    private List<RelationshipEntity> mapToRelationshipEntityList(RowSet<Row> rows) {
        List<RelationshipEntity> list = new ArrayList<>();
        for (Row row : rows) {
            list.add(mapToRelationshipEntity(row));
        }
        return list;
    }
}

File: apimgmt/repositories/relationship_repository.py
from sqlalchemy.orm import Session
import uuid
from typing import List, Optional, Union

from apimgmt.models.relationship import Relationship


class RelationshipRepository:
    def __init__(self, db: Session):
        self.db = db
    
    def create(self, relationship: Relationship) -> Relationship:
        """创建调用关系"""
        # 关系ID会在模型的默认值中设置，不需要在这里设置
        self.db.add(relationship)
        self.db.commit()
        self.db.refresh(relationship)
        return relationship
    
    def find_by_id(self, relationship_id: Union[uuid.UUID, str]) -> Optional[Relationship]:
        """根据ID查找调用关系"""
        return self.db.query(Relationship).filter(Relationship.id == relationship_id).first()
    
    def find_all(self) -> List[Relationship]:
        """查找所有调用关系"""
        return self.db.query(Relationship).all()
    
    def find_by_caller(self, caller_type: str, caller_id: Union[uuid.UUID, str]) -> List[Relationship]:
        """根据调用方查找调用关系"""
        return (
            self.db.query(Relationship)
            .filter(
                Relationship.caller_type == caller_type,
                Relationship.caller_id == caller_id
            )
            .all()
        )
    
    def find_by_callee(self, callee_type: str, callee_id: Union[uuid.UUID, str]) -> List[Relationship]:
        """根据被调用方查找调用关系"""
        return (
            self.db.query(Relationship)
            .filter(
                Relationship.callee_type == callee_type,
                Relationship.callee_id == callee_id
            )
            .all()
        )
    
    def update(self, relationship: Relationship) -> Optional[Relationship]:
        """更新调用关系"""
        existing_relationship = self.find_by_id(relationship.id)
        if existing_relationship:
            for key, value in relationship.__dict__.items():
                if key != '_sa_instance_state':
                    setattr(existing_relationship, key, value)
            self.db.commit()
            self.db.refresh(existing_relationship)
            return existing_relationship
        return None
    
    def delete(self, relationship_id: Union[uuid.UUID, str]) -> bool:
        """删除调用关系"""
        relationship = self.find_by_id(relationship_id)
        if relationship:
            self.db.delete(relationship)
            self.db.commit()
            return True
        return False
